File: modules/InMemory.Workers/InMemoryPoseWorker.cs
using System.Numerics;
using Engine.Client;
using Engine.Worker;
using InMemory;

namespace InMemory.Workers;

public partial class InMemoryPoseWorker : ComponentWorker<InMemoryPose>
{
    private Pose _pose = new Pose { Position = Vector3.Zero, Rotation = Quaternion.Identity };

    public Task<Pose> GetDataAsync(CancellationToken ct = default)
    {
        return Task.FromResult(_pose);
    }

    public Task SetDataAsync(Pose data, CancellationToken ct = default)
    {
        _pose = data;
        return Task.CompletedTask;
    }
}

File: modules/in-memory-workers/src/in-memory-pose-worker.ts
import { ComponentWorker } from "engine-worker";
import { encode, decode } from "@msgpack/msgpack";
import type { Pose } from "engine-client";

/**
 * In-memory worker for the IPose behaviour.
 * Stores position and rotation per entity.
 */
export class InMemoryPoseWorker extends ComponentWorker {
  readonly componentName = "InMemoryPose";
  readonly behaviourNames = ["IPose"];

  private pose: Pose = {
    position: { x: 0, y: 0, z: 0 },
    rotation: { x: 0, y: 0, z: 0, w: 1 },
  };

  async dispatch(
    _behaviourName: string,
    methodName: string,
    payload: Uint8Array,
  ): Promise<Uint8Array> {
    switch (methodName) {
      case "GetDataAsync": {
        return encode(this.pose);
      }
      case "SetDataAsync": {
        this.pose = decode(payload) as Pose;
        return new Uint8Array();
      }
      default:
        throw new Error(`Unknown method '${methodName}' on behaviour IPose`);
    }
  }
}
